### PR TITLE
Redo PR #24811 properly

### PR DESCRIPTION
**Pull request by hanna-kruppe**
Transplant the relevant changes (turns out to be all of them) to `grammar.md`, and remove all grammar talk from `reference.md`. Sorry for the chaos.

The second commit, further below, goes over the comments and whitespace sections.

r? @steveklabnik

### CODE REVIEW

File: src/doc/reference.md [hunk 1]
```diff
@@ -86,19 +81,10 @@ that does _not_ occur in the set of [keywords](#keywords).
...
-
 ## Comments
 
-Comments in Rust code follow the general C++ style of line and block-comment
-forms. Nested block comments are supported.
+Comments in Rust code follow the general C++ style of line (`//`) and
+block-comment (`/* ... */`) forms. Nested block comments are supported.
 
 Line comments beginning with exactly _three_ slashes (`///`), and block
 comments beginning with exactly one repeated asterisk in the block-open
```
Comment (mdinger) [inline src/doc/reference.md]: Does `/// Foo` really expand to `#[doc = "/// Foo"]`? @chris-morgan [states](http://chrismorgan.info/blog/rust-cfg_attr.html) it expands to `#[doc = " Foo"]`.

Comment (rkruppe) [inline src/doc/reference.md]: Good question. I see no reason why it Chris Morgan would be wrong. I just didn't think to try it out (honestly didn't even see the triple slash!). I'm currently in the middle of a build, so I can't even check...

Comment (rkruppe) [inline src/doc/reference.md]: Meh, `rustc --pretty` does not show any `#[doc]` attributes, it just regurgitates the triple-slash comments.

Comment (mdinger) [inline src/doc/reference.md]: I don't know. What it says seems kinda weird though. You could probably try manually inputting it like: `#[doc = "/// Foo"]` and see what rustdoc spits out. That might clarify it.

**Comment by rkruppe** [comment]
@steveklabnik re-r? See the above discussion with @mdinger

File: src/doc/reference.md [hunk 2]
```diff
@@ -86,40 +81,34 @@ that does _not_ occur in the set of [keywords](#keywords).
...
 sequence (`/**`), are interpreted as a special syntax for `doc`
 [attributes](#attributes). That is, they are equivalent to writing
 `#[doc="..."]` around the body of the comment (this includes the comment
```
Comment (mdinger) [inline src/doc/reference.md]: around the body of the comment (~~this includes the comment characters themselves~~, ~~i.e.~~[that is,] `/// Foo` turns into `#[doc=" Foo"]`)

File: src/doc/reference.md (hunk 1 above)
Comment (rkruppe) [inline src/doc/reference.md]: `#[doc = "/// Foo"]` results in rustdoc spitting out "/// Foo" as documentation text, so Chris Morgan is definitely right (and by extension, so is my change :)).

Comment (mdinger) [inline src/doc/reference.md]: Good.

Comment (steveklabnik) [inline src/doc/reference.md]: Yeah, this is purely an implementation detail for now, as far as I'm concerned.

**Pull request merged**